File: python/checkpreprocessresults.py
#!/usr/bin/python2
#
# python-mode indent C-c < or C-c >
# python-mode comment/uncomment region M-;
# Pick a random event directory and compare results
###########################################################################
# IMPORTS
###########################################################################
from obspy.core import read
from obspy.signal.invsim import cosTaper
import numpy as np
import scipy.fftpack as fft
from scipy.signal.signaltools import detrend
import random, os, os.path, re
import matplotlib.pyplot as plt
from preprocessor import calculate_slowness, detrend_taper_rotate, NoSlownessError


###########################################################################
# Get random event directory
###########################################################################
checkdir = '/media/TerraS/TEST'
stations = os.listdir(checkdir)
station = stations[random.randint(0,len(stations)-1)]
events = os.listdir(checkdir + '/' + station)
event = events[random.randint(0,len(events)-1)]
eventdir = os.path.join(checkdir,station,event)
print "Going to compare data in event directory: ", eventdir
###########################################################################
# Read in files
###########################################################################

reOld = re.compile(r'^(\d{4}\.\d{3}\.\d{2}\.\d{2}\.\d{2})\.\d{4}\.(\w{2})\.(\w{4})\.\.(\w{3}).*')
reNew = re.compile(r'^STACK_(\w)\.sac')

fs = os.listdir(eventdir)
newfs = []
oldfs = []
comps = []

###########################################################################
# Checking processing function on random directories:
###########################################################################
for f in fs:
    try:
        m1 = reOld.match(f)
        comps.append((m1.group(4),f)) # Save the component extension, see Regex above.
    except AttributeError as e:
        print "No match on file:",f
if len(comps) != 3:
    print "Did not register 3 components in directory:", eventdir
    # Sort in decending alphabetical, so 'E' is [0] 'N' is [1] and 'Z' is [2]
    # Pull out sacfiles from zipped sorted list.
comps.sort()
_ , sacfiles = zip(*comps)
try:
    calculate_slowness(eventdir, sacfiles)
except NoSlownessError:
    print "did not find a good slowness in event folder:", eventdir
#detrend_taper_rotate(eventdir,sacfiles)




# for f in fs:
#     try:
#         m1 = reOld.match(f)
#         if m1 is not None:
#             oldfs.append(eventdir + '/' + m1.group(0))
#         m2 = reNew.match(f)
#         if m2 is not None:
#             newfs.append(eventdir + '/' + m2.group(0))
#         #comps.append((m1.group(4),fs)) # Save the component extension, see Regex above.
#     except AttributeError:
#         pass

# trOld = []
# trNew = []

# for f in sorted(oldfs):
#     st = read(f)
#     st[0].data = detrend(st[0].data)
#     ctap = cosTaper(len(st[0].data))
#     st[0].data = st[0].data * ctap
#     trOld.append(st[0]) 
    
# for f in sorted(oldfs):
#     st = read(f)
#     trNew.append(st[0]) 

# # Plot and compare
# N = len(trNew[1].data)
# t = range(N)
# dt = trNew[0].stats.delta

 
#plt.subplot(2,1,1)
#
#plt.title('Frequency comparison')
#plt.subplot(2,1,2)
#
#plt.show()

File: matlab/checkpreprocessresults.m
%% Get random event directory
clear all; close all;
addpath functions
addpath sac
checkdir = '/media/TerraS/X5';
station = 'CRLN';
%checkdir = '/media/TerraS/CNSN';
%station = 'ULM';
events = dir([checkdir ,'/' , station]);
while true
    event = events(randi(length(events))).name;
    if str2num(event)
          break
    end
end
eventdir = fullfile(checkdir,station,event);
fprintf('Going to compare data in event directory: %s\n', eventdir)

%% Read in files

fs = dir(eventdir);
ind = 1;
for f = fs'
    if strfind(f.name,'stack_R')
        header = readsac(fullfile(eventdir,f.name));
        [~,rcomp] = readsac(fullfile(eventdir,f.name));
        
    end
    if strfind(f.name,'stack_Z')
        [~,zcomp] = readsac(fullfile(eventdir,f.name));
    end
    if strfind(f.name,'BHZ')
        [~,Zcomp] = readsac(fullfile(eventdir,f.name));
    end
end

%[fname,I] = sort(fname);
%header = header(I);
%comp = comp(I);

%fprintf('\ndata in comp variable with array order: \n');
%for f = fname
%    fprintf('%s\n',cell2mat(f))
%end

dt = header.DELTA;
pslow = header.USER0;

[p,s] = freetran(rcomp',zcomp',pslow,6.06,3.5,1);

Zcomp = Zcomp(1:16384);
Zcomp = Zcomp./max(Zcomp);

t = 1:length(p);

t0 = round( (header.T0 - header.B) /dt );
t4 = round( (header.T4 - header.B)/dt );
t7 = round( (header.T7 - header.B)/dt );

figure(34)
subplot(2,1,1)
plot(t,p./max(p))
xlim([ t0 - 200  , t0 + 2000])
line([ t0; t0], [-1; 1], ...
    'LineWidth', 2, 'Color', [.8 .8 .2]);
line([ t4; t4], [ -1, 1], ...
    'LineWidth', 2, 'Color', [.8 .2 .8]);
line([ t7; t7], [ -1, 1], ...
    'LineWidth', 2, 'Color', [.8 .4 .4]);
%line([ t3; t3], [ -1, 1], ...
%    'LineWidth', 2, 'Color', [.4 .4 .4]);
title(sprintf('Depth is %f',header.EVDP))
legend('PTrace','P','pP','PP')
subplot(2,1,2)
plot(zcomp./max(zcomp))
xlim([ t0 - 200  , t0 + 2000])



